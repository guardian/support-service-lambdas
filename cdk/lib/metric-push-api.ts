import { GuApiLambda } from '@guardian/cdk';
import { GuAlarm } from '@guardian/cdk/lib/constructs/cloudwatch';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuPutCloudwatchMetricsPolicy } from '@guardian/cdk/lib/constructs/iam';
import { type App, Duration } from 'aws-cdk-lib';
import { CfnBasePathMapping, CfnDomainName } from 'aws-cdk-lib/aws-apigateway';
import {
	ComparisonOperator,
	Metric,
	TreatMissingData,
} from 'aws-cdk-lib/aws-cloudwatch';
import { LoggingFormat } from 'aws-cdk-lib/aws-lambda';
import { CfnRecordSet } from 'aws-cdk-lib/aws-route53';
import { nodeVersion } from './node-version';

export class MetricPushApi extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);
		const app = 'metric-push-api';
		const nameWithStage = `${app}-${this.stage}`;

		// Lambda & API Gateway
		const commonEnvironmentVariables = {
			App: app,
			Stack: this.stack,
			Stage: this.stage,
		};

		const lambda = new GuApiLambda(this, `${app}-lambda`, {
			description:
				'API triggered lambda to push a metric to cloudwatch so we can alarm on errors',
			functionName: nameWithStage,
			loggingFormat: LoggingFormat.TEXT,
			fileName: `${app}.zip`,
			handler: 'index.handler',
			runtime: nodeVersion,
			memorySize: 512,
			timeout: Duration.seconds(60),
			environment: commonEnvironmentVariables,
			// Create an alarm
			monitoringConfiguration: {
				noMonitoring: true,
			},
			app: app,
			api: {
				id: nameWithStage,
				restApiName: nameWithStage,
				description: `API Gateway endpoint for the ${nameWithStage} lambda`,
				proxy: true,
				deployOptions: {
					stageName: this.stage,
				},
			},
		});

		const cloudwatchPutMetricPolicy = new GuPutCloudwatchMetricsPolicy(this);
		lambda.role?.attachInlinePolicy(cloudwatchPutMetricPolicy);

		// DNS
		const domainName = new CfnDomainName(this, 'MetricPushDomainName', {
			regionalCertificateArn: `arn:aws:acm:${this.region}:${this.account}:certificate/b384a6a0-2f54-4874-b99b-96eeff96c009`,
			domainName: `metric-push-api-${this.stage.toLowerCase()}.support.guardianapis.com`,
			endpointConfiguration: {
				types: ['REGIONAL'],
			},
		});

		new CfnBasePathMapping(this, 'MetricPushBasePathMapping', {
			restApiId: lambda.api.restApiId,
			domainName: domainName.ref,
			stage: lambda.api.deploymentStage.stageName,
		});

		const dnsRecord = new CfnRecordSet(this, 'MetricPushDNSRecord', {
			name: `metric-push-api-${this.stage.toLowerCase()}.support.guardianapis.com`,
			type: 'CNAME',
			comment: `CNAME for metric-push-api API ${this.stage}`,
			hostedZoneName: 'support.guardianapis.com.',
			ttl: '120',
			resourceRecords: [domainName.attrRegionalDomainName],
		});
		dnsRecord.overrideLogicalId('MetricPushDNSRecord');

		// Alarms
		new GuAlarm(this, '5xxApiAlarm', {
			app,
			alarmName: `URGENT 9-5 - ${this.stage} ${nameWithStage} API Gateway is returning 5XX errors`,
			threshold: 2,
			evaluationPeriods: 1,
			snsTopicName: `alarms-handler-topic-${this.stage}`,
			actionsEnabled: this.stage === 'PROD',
			treatMissingData: TreatMissingData.NOT_BREACHING,
			comparisonOperator: ComparisonOperator.GREATER_THAN_THRESHOLD,
			metric: new Metric({
				metricName: '5XXError',
				namespace: 'AWS/ApiGateway',
				statistic: 'Sum',
				period: Duration.seconds(60),
				dimensionsMap: {
					ApiName: nameWithStage,
				},
			}),
		});

		new GuAlarm(this, 'HighClientSideErrorRateAlarm', {
			app,
			alarmName: `URGENT 9-5 - ${this.stage} fatal client-side errors are being reported to sentry for support-frontend`,
			alarmDescription: `Impact - some or all browsers are failing to render support client side pages. Log in to Sentry to see these errors: https://the-guardian.sentry.io/discover/results/?project=1213654&query="Fatal error rendering page"&queryDataset=error-events&sort=-count&statsPeriod=24h Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit ${nameWithStage}`,
			threshold: 2,
			evaluationPeriods: 5,
			datapointsToAlarm: 3,
			snsTopicName: `alarms-handler-topic-${this.stage}`,
			actionsEnabled: this.stage === 'PROD',
			treatMissingData: TreatMissingData.NOT_BREACHING,
			comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
			metric: new Metric({
				metricName: 'metric-push-api-client-side-error',
				namespace: 'support-service-lambdas',
				statistic: 'Sum',
				period: Duration.seconds(60),
				dimensionsMap: {
					Stage: this.stage,
					App: app,
				},
			}),
		});
	}
}
