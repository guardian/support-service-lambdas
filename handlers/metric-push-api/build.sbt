// "Any .sbt files in foo, say foo/build.sbt, will be merged with the build definition for the entire build, but scoped to the hello-foo project."
// https://www.scala-sbt.org/0.13/docs/Multi-Project.html
name := "metric-push-api"
description:= "HTTP API to push a metric to cloudwatch so we can alarm on errors"

scalacOptions += "-Ypartial-unification"

assemblyJarName := "metric-push-api.jar"
riffRaffPackageType := assembly.value
riffRaffUploadArtifactBucket := Option("riffraff-artifact")
riffRaffUploadManifestBucket := Option("riffraff-builds")
riffRaffManifestProjectName := "MemSub::Membership Admin::Metric Push API"
riffRaffArtifactResources += (file("handlers/metric-push-api/cfn.yaml"), "cfn/cfn.yaml")

libraryDependencies ++= Seq()
