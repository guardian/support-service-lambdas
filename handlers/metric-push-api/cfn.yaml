AWSTemplateFormatVersion: "2010-09-09"
Description: HTTP API to push a metric to cloudwatch so we can alarm on errors

Parameters:
    Stage:
        Description: Stage name
        Type: String
        AllowedValues:
            - PROD
            - CODE

Conditions:
  CreateProdMonitoring: !Equals [ !Ref Stage, PROD ]

Mappings:
    StageMap:
        CODE:
            ApiName: metric-push-api-api-CODE
            DomainName: metric-push-api-code.support.guardianapis.com
        PROD:
            ApiName: metric-push-api-api-PROD
            DomainName: metric-push-api-prod.support.guardianapis.com
    Constants:
      Alarm:
        Process: Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit
        Urgent: URGENT 9-5 -

Resources:

    MetricPushProxyResource:
        Type: AWS::ApiGateway::Resource
        Properties:
            RestApiId: !Ref MetricPushAPI
            ParentId: !GetAtt [MetricPushAPI, RootResourceId]
            PathPart: metric-push-api
        DependsOn: MetricPushAPI

    MetricPushMethod:
        Type: AWS::ApiGateway::Method
        Properties:
            AuthorizationType: NONE
            ApiKeyRequired: false
            RestApiId: !Ref MetricPushAPI
            ResourceId: !Ref MetricPushProxyResource
            HttpMethod: GET
            Integration:
              Type: MOCK
              RequestTemplates:
                application/json: '{"statusCode": 200}'
              IntegrationResponses:
                - StatusCode: 204
                  ResponseParameters:
                    method.response.header.Cache-control: '''no-cache'''
            MethodResponses:
              - StatusCode: 204
                ResponseParameters:
                  method.response.header.Cache-control: true
        DependsOn:
        - MetricPushAPI
        - MetricPushProxyResource

    MetricPushAPI:
        Type: "AWS::ApiGateway::RestApi"
        Properties:
            Description: HTTP API to push a metric to cloudwatch so we can alarm on errors
            Name: !FindInMap [StageMap, !Ref Stage, ApiName]

    MetricPushAPIStage:
        Type: AWS::ApiGateway::Stage
        Properties:
            Description: Stage for metric-push-api
            RestApiId: !Ref MetricPushAPI
            DeploymentId: !Ref MetricPushAPIDeployment
            StageName: !Sub ${Stage}
            MethodSettings:
            - ResourcePath: '/*'
              HttpMethod: '*'
              LoggingLevel: ERROR
              DataTraceEnabled: true
        DependsOn: MetricPushMethod

    MetricPushAPIDeployment:
        Type: AWS::ApiGateway::Deployment
        Properties:
            Description: Deploys metric-push-api into an environment/stage
            RestApiId: !Ref MetricPushAPI
        DependsOn: MetricPushMethod

    MetricPushDomainName:
      Type: "AWS::ApiGateway::DomainName"
      Properties:
        RegionalCertificateArn: # only for *.support.guardianapis.com
          !Sub arn:aws:acm:${AWS::Region}:${AWS::AccountId}:certificate/b384a6a0-2f54-4874-b99b-96eeff96c009
        DomainName: !FindInMap [ StageMap, !Ref Stage, DomainName ]
        EndpointConfiguration:
          Types:
            - REGIONAL

    MetricPushBasePathMapping:
      Type: "AWS::ApiGateway::BasePathMapping"
      Properties:
        RestApiId: !Ref MetricPushAPI
        DomainName: !Ref MetricPushDomainName
        Stage: !Sub ${Stage}
      DependsOn:
      - MetricPushAPI
      - MetricPushAPIStage
      - MetricPushDomainName

    MetricPushDNSRecord:
      Type: AWS::Route53::RecordSet
      Properties:
        HostedZoneName: support.guardianapis.com.
        Name: !FindInMap [ StageMap, !Ref Stage, DomainName ]
        Comment: !Sub CNAME for metric-push-api API ${Stage}
        Type: CNAME
        TTL: '120'
        ResourceRecords:
        - !GetAtt [ MetricPushDomainName, RegionalDomainName ]
      DependsOn:
        - MetricPushDomainName

    5xxApiAlarm:
      Type: AWS::CloudWatch::Alarm
      Condition: CreateProdMonitoring
      Properties:
        AlarmActions:
          - !Sub arn:aws:sns:${AWS::Region}:${AWS::AccountId}:reader-revenue-dev
        AlarmName: !Join
        - ' '
        - - !FindInMap [ Constants, Alarm, Urgent ]
          - !Ref 'Stage'
          - !FindInMap [StageMap, !Ref Stage, ApiName]
          - 'API Gateway is returning 5XX errors'
        ComparisonOperator: GreaterThanThreshold
        Dimensions:
          - Name: ApiName
            Value: !FindInMap [StageMap, !Ref Stage, ApiName]
          - Name: Stage
            Value: !Sub ${Stage}
        EvaluationPeriods: 1
        MetricName: 5XXError
        Namespace: AWS/ApiGateway
        Period: 60
        Statistic: Sum
        Threshold: 2
        TreatMissingData: notBreaching

    HighClientSideErrorRateAlarm:
      Type: AWS::CloudWatch::Alarm
      Condition: CreateProdMonitoring
      Properties:
        AlarmActions:
          - !Sub arn:aws:sns:${AWS::Region}:${AWS::AccountId}:reader-revenue-dev
        AlarmName: !Join
          - ' '
          - - !FindInMap [ Constants, Alarm, Urgent ]
            - !Ref 'Stage'
            - 'client-side fatal errors are being reported to sentry for support-frontend'
        AlarmDescription: !Join
          - ' '
          - - 'Impact - some or all browsers are failing to render support client side pages. Log in to Sentry.io to investigate.'
            - !FindInMap [ Constants, Alarm, Process ]
            - !FindInMap [StageMap, !Ref Stage, ApiName]
        MetricName: Count
        Namespace: AWS/ApiGateway
        Dimensions:
          - Name: ApiName
            Value: !FindInMap [StageMap, !Ref Stage, ApiName]
        ComparisonOperator: GreaterThanOrEqualToThreshold
        Threshold: 2
        DatapointsToAlarm: 3
        Period: 60
        EvaluationPeriods: 5
        Statistic: Sum
        TreatMissingData: notBreaching
      DependsOn:
        - MetricPushAPI
