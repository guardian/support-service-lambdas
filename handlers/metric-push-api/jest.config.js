/** @type {import('ts-jest').JestConfigWithTsJest} */
module.exports = {
	preset: 'ts-jest',
	testEnvironment: 'node',
	runner: 'groups',
	clearMocks: true,
	moduleNameMapper: {
		'@modules/(.*)/(.*)$': '<rootDir>/../../modules/$1/src/$2',
		'@modules/(.*)$': '<rootDir>/../../modules/$1',
	},
};
