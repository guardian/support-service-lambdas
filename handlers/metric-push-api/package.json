{
	"name": "metric-push-api",
	"scripts": {
		"type-check": "tsc --noEmit",
		"build": "esbuild --bundle --platform=node --target=node18 --outfile=target/index.js src/index.ts",
		"lint": "eslint src/**/*.ts",
		"package": "pnpm type-check && pnpm lint && pnpm check-formatting && pnpm build && cd target && zip -qr metric-push-api.zip ./*.js",
		"check-formatting": "prettier --check **.ts",
		"fix-formatting": "prettier --write **.ts"
	},
	"dependencies": {},
	"devDependencies": {
		"@types/aws-lambda": "^8.10.147"
	}
}
