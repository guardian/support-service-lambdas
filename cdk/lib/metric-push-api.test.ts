import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MetricPushApi } from './metric-push-api';

describe('The MetricPushApi stack', () => {
	it('matches the snapshot', () => {
		const app = new App();

		const codeStack = new MetricPushApi(app, 'CODE');
		const prodStack = new MetricPushApi(app, 'PROD');

		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
