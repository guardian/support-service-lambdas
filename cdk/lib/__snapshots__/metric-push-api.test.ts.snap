// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MetricPushApi stack matches the snapshot 1`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "HTTP API to push a metric to cloudwatch so we can alarm on errors",
  "Mappings": {
    "Constants": {
      "Alarm": {
        "Process": "Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit",
        "Urgent": "URGENT 9-5 -",
      },
    },
    "StageMap": {
      "CODE": {
        "AlarmActionsEnabled": false,
        "ApiName": "metric-push-api-api-CODE",
        "DomainName": "metric-push-api-code.support.guardianapis.com",
      },
      "PROD": {
        "AlarmActionsEnabled": true,
        "ApiName": "metric-push-api-api-PROD",
        "DomainName": "metric-push-api-prod.support.guardianapis.com",
      },
    },
  },
  "Metadata": {
    "gu:cdk:constructs": [
      "GuAlarm",
      "GuAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "Stage": {
      "AllowedValues": [
        "PROD",
        "CODE",
      ],
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": {
    "5xxApiAlarmF6ACE36F": {
      "Properties": {
        "ActionsEnabled": false,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":alarms-handler-topic-CODE",
              ],
            ],
          },
        ],
        "AlarmName": "URGENT 9-5 - CODE metric-push-api-CODE API Gateway is returning 5XX errors",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "metric-push-api-api-CODE",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "5XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "HighClientSideErrorRateAlarm0026FCB2": {
      "Properties": {
        "ActionsEnabled": false,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":alarms-handler-topic-CODE",
              ],
            ],
          },
        ],
        "AlarmDescription": "Impact - some or all browsers are failing to render support client side pages. Log in to Sentry to see these errors: https://the-guardian.sentry.io/discover/results/?project=1213654&query="Fatal error rendering page"&queryDataset=error-events&sort=-count&statsPeriod=24h Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit metric-push-api-CODE",
        "AlarmName": "URGENT 9-5 - CODE fatal client-side errors are being reported to sentry for support-frontend",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 5,
        "Metrics": [
          {
            "Expression": "mtotalcount - (m5xxcount + m4xxcount)",
            "Id": "expr_1",
            "Label": "mtotal2xx",
          },
          {
            "Id": "mtotalcount",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "ApiName",
                    "Value": "metric-push-api-api-CODE",
                  },
                ],
                "MetricName": "Count",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m5xxcount",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "ApiName",
                    "Value": "metric-push-api-api-CODE",
                  },
                ],
                "MetricName": "5XXError",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m4xxcount",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "ApiName",
                    "Value": "metric-push-api-api-CODE",
                  },
                ],
                "MetricName": "4XXError",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MetricPushAPI": {
      "Properties": {
        "Description": "HTTP API to push a metric to cloudwatch so we can alarm on errors",
        "Name": {
          "Fn::FindInMap": [
            "StageMap",
            {
              "Ref": "Stage",
            },
            "ApiName",
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "MetricPushAPIDeployment": {
      "DependsOn": [
        "MetricPushMethod",
      ],
      "Properties": {
        "Description": "Deploys metric-push-api into an environment/stage",
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "MetricPushAPIStage": {
      "DependsOn": [
        "MetricPushMethod",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "MetricPushAPIDeployment",
        },
        "Description": "Stage for metric-push-api",
        "MethodSettings": [
          {
            "DataTraceEnabled": true,
            "HttpMethod": "*",
            "LoggingLevel": "ERROR",
            "ResourcePath": "/*",
          },
        ],
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
        "StageName": {
          "Fn::Sub": "\${Stage}",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "MetricPushBasePathMapping": {
      "Properties": {
        "DomainName": {
          "Ref": "MetricPushDomainName",
        },
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
        "Stage": "CODE",
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "MetricPushDNSRecord": {
      "Properties": {
        "Comment": "CNAME for metric-push-api API CODE",
        "HostedZoneName": "support.guardianapis.com.",
        "Name": "metric-push-api-code.support.guardianapis.com",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "MetricPushDomainName",
              "RegionalDomainName",
            ],
          },
        ],
        "TTL": "120",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "MetricPushDomainName": {
      "Properties": {
        "DomainName": "metric-push-api-code.support.guardianapis.com",
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:acm:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":certificate/b384a6a0-2f54-4874-b99b-96eeff96c009",
            ],
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "MetricPushMethod": {
      "DependsOn": [
        "MetricPushAPI",
        "MetricPushProxyResource",
      ],
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Cache-control": "'no-cache'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{"statusCode": 200}",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Cache-control": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Ref": "MetricPushProxyResource",
        },
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "MetricPushProxyResource": {
      "DependsOn": [
        "MetricPushAPI",
      ],
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "MetricPushAPI",
            "RootResourceId",
          ],
        },
        "PathPart": "metric-push-api",
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
  },
}
`;

exports[`The MetricPushApi stack matches the snapshot 2`] = `
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Description": "HTTP API to push a metric to cloudwatch so we can alarm on errors",
  "Mappings": {
    "Constants": {
      "Alarm": {
        "Process": "Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit",
        "Urgent": "URGENT 9-5 -",
      },
    },
    "StageMap": {
      "CODE": {
        "AlarmActionsEnabled": false,
        "ApiName": "metric-push-api-api-CODE",
        "DomainName": "metric-push-api-code.support.guardianapis.com",
      },
      "PROD": {
        "AlarmActionsEnabled": true,
        "ApiName": "metric-push-api-api-PROD",
        "DomainName": "metric-push-api-prod.support.guardianapis.com",
      },
    },
  },
  "Metadata": {
    "gu:cdk:constructs": [
      "GuAlarm",
      "GuAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "Stage": {
      "AllowedValues": [
        "PROD",
        "CODE",
      ],
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": {
    "5xxApiAlarmF6ACE36F": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":alarms-handler-topic-PROD",
              ],
            ],
          },
        ],
        "AlarmName": "URGENT 9-5 - PROD metric-push-api-PROD API Gateway is returning 5XX errors",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "metric-push-api-api-PROD",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "5XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "HighClientSideErrorRateAlarm0026FCB2": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":alarms-handler-topic-PROD",
              ],
            ],
          },
        ],
        "AlarmDescription": "Impact - some or all browsers are failing to render support client side pages. Log in to Sentry to see these errors: https://the-guardian.sentry.io/discover/results/?project=1213654&query="Fatal error rendering page"&queryDataset=error-events&sort=-count&statsPeriod=24h Follow the process in https://docs.google.com/document/d/1_3El3cly9d7u_jPgTcRjLxmdG2e919zCLvmcFCLOYAk/edit metric-push-api-PROD",
        "AlarmName": "URGENT 9-5 - PROD fatal client-side errors are being reported to sentry for support-frontend",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 5,
        "Metrics": [
          {
            "Expression": "mtotalcount - (m5xxcount + m4xxcount)",
            "Id": "expr_1",
            "Label": "mtotal2xx",
          },
          {
            "Id": "mtotalcount",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "ApiName",
                    "Value": "metric-push-api-api-PROD",
                  },
                ],
                "MetricName": "Count",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m5xxcount",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "ApiName",
                    "Value": "metric-push-api-api-PROD",
                  },
                ],
                "MetricName": "5XXError",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m4xxcount",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "ApiName",
                    "Value": "metric-push-api-api-PROD",
                  },
                ],
                "MetricName": "4XXError",
                "Namespace": "AWS/ApiGateway",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 2,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MetricPushAPI": {
      "Properties": {
        "Description": "HTTP API to push a metric to cloudwatch so we can alarm on errors",
        "Name": {
          "Fn::FindInMap": [
            "StageMap",
            {
              "Ref": "Stage",
            },
            "ApiName",
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "MetricPushAPIDeployment": {
      "DependsOn": [
        "MetricPushMethod",
      ],
      "Properties": {
        "Description": "Deploys metric-push-api into an environment/stage",
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "MetricPushAPIStage": {
      "DependsOn": [
        "MetricPushMethod",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "MetricPushAPIDeployment",
        },
        "Description": "Stage for metric-push-api",
        "MethodSettings": [
          {
            "DataTraceEnabled": true,
            "HttpMethod": "*",
            "LoggingLevel": "ERROR",
            "ResourcePath": "/*",
          },
        ],
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
        "StageName": {
          "Fn::Sub": "\${Stage}",
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "MetricPushBasePathMapping": {
      "Properties": {
        "DomainName": {
          "Ref": "MetricPushDomainName",
        },
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
        "Stage": "PROD",
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "MetricPushDNSRecord": {
      "Properties": {
        "Comment": "CNAME for metric-push-api API PROD",
        "HostedZoneName": "support.guardianapis.com.",
        "Name": "metric-push-api-prod.support.guardianapis.com",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "MetricPushDomainName",
              "RegionalDomainName",
            ],
          },
        ],
        "TTL": "120",
        "Type": "CNAME",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "MetricPushDomainName": {
      "Properties": {
        "DomainName": "metric-push-api-prod.support.guardianapis.com",
        "EndpointConfiguration": {
          "Types": [
            "REGIONAL",
          ],
        },
        "RegionalCertificateArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:acm:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":certificate/b384a6a0-2f54-4874-b99b-96eeff96c009",
            ],
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/support-service-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "MetricPushMethod": {
      "DependsOn": [
        "MetricPushAPI",
        "MetricPushProxyResource",
      ],
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Cache-control": "'no-cache'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": {
            "application/json": "{"statusCode": 200}",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Cache-control": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": {
          "Ref": "MetricPushProxyResource",
        },
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "MetricPushProxyResource": {
      "DependsOn": [
        "MetricPushAPI",
      ],
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "MetricPushAPI",
            "RootResourceId",
          ],
        },
        "PathPart": "metric-push-api",
        "RestApiId": {
          "Ref": "MetricPushAPI",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
  },
}
`;
